### DISCUSSION TITLE
Dependent validators

### DISCUSSION THREAD

**Discussion by antonagestam** [pydantic/pydantic#2959] [Feature Request]
Hi, and thanks for a great library 👋

I'd like to propose adding more advanced options for validating dependent fields.

The use case I have is quite simple to describe: I want to create a model that has the ability to be passed domestic phone numbers and for them to be normalized to the correct format for the provided country code. I can get pretty close with the current set of facilities:

```python
class User(BaseModel):
    country: CountryCode
    phone_number: FormattedPhoneNumber

    @validator("phone_number", pre=True)
    def valid_phone_number_for_country(cls, value, values, **kwargs):
        if "country" not in values:
            return None
        return normalize_phone_number(value, values["country"])
```

The problem with this code is that `values["country"]` is passed to `normalize_phone_number()` without first being validated to be a `CountryCode`. One way to solve for that would be to parse it with something like `assert isinstance(country, CountryCode)` within `valid_phone_number_for_country()`. But that has a few issues:

- The `country` field is now being doubled parsed.
- This is "shotgun parsing" e.g. the two validations are likely to drift and become out of sync, especially when considering how `valid_phone_number_for_country` would also be responsible for providing the correct error message for `country`.
- It arguably breaks the single responsibility principle with regards to `valid_phone_number_for_country` now doing two things.

Without giving concern to feasibility of implementation, performance implications etc, my naive approach to address is that I'd like to have an API that allows validators to be executed in-between `pre=True` and `pre=False`. Ideally this would also allow working with already validated fields in a typed way. So an API I can think of that would satisfy this would be along these lines:

```python
class User(BaseModel):
    country: CountryCode
    phone_number: FormattedPhoneNumber

    @validator("phone_number", post=["country"])
    def valid_phone_number_for_country(cls, value: object, country: CountryCode) -> FormattedPhoneNumber:
        return normalize_phone_number(value, country)
```

The expectation here would be that `valid_phone_number_for_country` is executed after `country` is proven to be a valid `CountryCode`, but before default validation for `phone_number` is executed.

I am guessing that this could be a bit challenging to implement as it would require constructing a graph for the field validation order to be correct. In addition to that, behavior would have to be explored for when that graph isn't resolvable. I think ideally that should give errors at import-time.

What do you think?

Cheers ☀️

#### Edit

This is related to, but different from https://github.com/samuelcolvin/pydantic/issues/1134. Also related to https://github.com/samuelcolvin/pydantic/issues/483.

**Comment by jordantshaw** [comment]
Great suggestion, this could be a very useful feature.  I would be interested in using something like this as well

**Comment by sblack-usu** [comment]
Isn't this accomplished with [field ordering](https://pydantic-docs.helpmanual.io/usage/models/#field-ordering)?  In your example, the country would be validated before the phone number because of the ordering.

From the field ordering documentation:
```validation is performed in the order fields are defined; fields validators can access the values of earlier fields, but not later ones```

**Reply by jordantshaw** [reply]
@sblack-usu You're correct that validation occurs based upon field ordering.  However, what if you wanted to order your fields in alphabetical order, but you have fieldA that relies on fieldB for its validation criteria.  In this case you would have to order fieldB before fieldA, and you would not be able to return the data in alphabetical order.

**Reply by sblack-usu** [reply]
I am only pointing out that there is a solution that is already implemented and available for @antonagestam today.

**Reply by antonagestam** [reply]
No, field ordering only partly solves my issue, but is definitely a lot better situation than what I had gathered. The issue is about having a typed API and statically typed access to already parsed values. Given how the field parsing order works, I think my proposed API should be a lot easier to achieve since it means no graph needs to be constructed. Instead a validator could yield errors at import-time if its dependencies violates the ordering.
